src/main/java/duke/TaskList.java

```diff
@@ -0,0 +1,167 @@
+     */
+    private TaskList(ArrayList<Task> oldTasks) {
+        tasks = new ArrayList<Task>(oldTasks);
+    }
+    /**
+     * Get the total number of tasks in the list.
+     *
+     * @return sizeOfTasks
+     */
+    int numberOfTasks() {
+        return tasks.size();
+    }
+    /**
+     * Display to the user in a pretty format all the tasks in the list,
+     * starting from 1.
+     *
+     * @return TaskList<T>
+     *
+     */
+    TaskList<T> listAllTasks() {
+        AtomicLong counter = new AtomicLong();
+        tasks.stream().forEach(task ->
+                System.out.println(counter.incrementAndGet() + "." + task));
+        /*
+        for (int i=0; i < numberOfTasks(); i++) {
+            System.out.println(i + INCREMENT + "." + get(i));
+        }
+        */
+        return new TaskList<T>(tasks);
+    }
+
+    <T extends Task> TaskList<T> listFindTasks(String find, TaskList<T> list) {
+        AtomicLong counter = new AtomicLong();
+        tasks.stream().filter(task ->
+                task.getDescription().contains(find.substring(1))).forEach(task ->
+                        System.out.println(counter.incrementAndGet() + "." + task));
+        /*
+        for (int i=0; i < numberOfTasks(); i++) {
+            Task currTask = (Task)get(i);
+            if (currTask.getDescription().contains(find.substring(1))) {
+                System.out.println(i + INCREMENT + "." + currTask);
+            }
+        }
+        */
+        return new TaskList<T>(tasks);
+    }
+
+    /**
+     * Get a task in a specific position
+     * @return positionOfTask
+     */
+    Task get(int index) {
+        return tasks.get(index);
+    }
+    /**
+     * Insert a new Task. However, due to the TaskList being immutable, a
+     * new TaskList is returned with the added task.
+     *
+     * @param newTask Task to add
+     * @return TaskList<T>
+     */
+    TaskList<T> add(Task newTask) {
+        TaskList<T> newTasks = new TaskList<T>(tasks);
+        newTasks.tasks.add(newTask);
+        System.out.println("SUI I have added this task!: " + newTask);
+        return newTasks;
+    }
+
+    TaskList<T> addMultipleTasks(Task... newTask) {
+        TaskList<T> newTasks = new TaskList<T>(tasks);
+        Arrays.stream(newTask).forEach(task -> {
+            newTasks.tasks.add(task);
+            System.out.println("SUI I have added this task!: " + task);
+        });
+        /*
+        for (Task task: newTask) {
+            newTasks.tasks.add(task);
+            System.out.println("SUI I have added this task!: " + task);
+        }
+        */
+        return newTasks;
+    }
+
+    /**
+     * Modify a task in the list, such as to mark it as done or undone. Due
+     * to the immutability nature of TaskList, a new TaskList will be
+     * returned with the modified marked / unmarked status of a particular
+     * task.
+     *
+     * @return TaskList<T>
+     */
+    TaskList<T> set(int index, Task task) {
```

Comment:
setTaskStatus would be a more appropriate name in my opinion
```suggestion
    TaskList<T> setTaskStatus(int index, Task task) {
```